### PR TITLE
WIP: Ezjsonm support.

### PR DESCRIPTION
**Pull request by mefyl**
This add support for `[%ezjsonm ...]` expressions. The modifications to the expression expander are pretty straightforward. I didn't support patterns yet as I want to test the water first before committing any further.

More question arise on the naming side.
* Firstly, the library itself is named after Yojson. This is probably not *that* problematic and may be seen as an artifact of the past to preserve backward compatibility, even if it supports other formats.
* I combined all expanders in a single ppx driver. We may want to split them, however the current driver is `ppx_yojson`, so adding `ppx_yojson.ezjsonm` looks a bit asymmetrical and weird. Also if we split them, one could argue that we'd want to use `[%json]` and pick the driver we want. Maybe we could have `ppx_yojson` that has all drivers with explicit `%yojson` and `%ezjsonm` extensions, and specific `ppx_yojson.yojson` and `ppx_yojson.ezjsonm` that define `%json` extensions ?

### CODE REVIEW

**Comment by mefyl** [comment]
I'm not sure about the CI error, it seems to use 4.12, which ocamlformat 0.15 is not compatible with. Should I bump ocamlformat to 0.17 ?

**Comment by NathanReb** [comment]
Ouh yeah, let me update ocamlformat separately and then you can rebase!

**Comment by mefyl** [comment]
Any idea regarding the naming questions ? Any other feedback on the PR ? :-)

**Comment by hhugo** [comment]
Should this define `[%jsonm ...]` instead ? Ezjsonm is more or less an helper library on top of jsonm

**Comment by NathanReb** [comment]
> Should this define `[%jsonm ...]` instead ? Ezjsonm is more or less an helper library on top of jsonm

It is but the types from the generated value are defined by ezjsonm, there no producer or consumer of those in jsonm IIRC.

@mefyl Sorry I haven't had the time to look at this but will do ASAP!

**Comment by mefyl** [comment]
Indeed, AFAICT jsonm itself doesn't define a datatype, "just" lexemes as it is pretty low-level.

@NathanReb no problem !

**Review by NathanReb — COMMENTED** [review]
This looks good! Thanks for working on this, it's definitely a valuable addition!

Since there's no type involved I probably would have used a record instead of a functor but it looks good as it is I can always refactor it later if I really really want to avoid the functor.

Do you plan on adding support for patterns as well? If not, I'm happy to take care of it so don't sweat it!

Could you add a changelog entry about this?

I just have a single request about the test setup (see below) but otherwise its good to go.

About the extension naming thing, the current situation seems fine to me, i.e. having `[%yojson ...]` and `[%ezjsonm ...]`. It's probably a good idea to re-release this as `ppx_json` in the future but I think it's fine to release both under `ppx_yojson` for now!

File: test/rewriter/ezjsonm.mli [hunk 1]
```diff
@@ -0,0 +1 @@
+
```
[the remark this answers is not in the document]
Comment (mefyl) [reply test/rewriter/ezjsonm.mli:2]: Certainly, done

**Comment by mefyl** [comment]
The workflow seems to consistently fail in yojson and I don't think I touched anything there. But let me know if I'm mistaken.

**Comment by NathanReb** [comment]
From a very quick look it seems that's the new dune default targets that builds the error test cases and therefore fails. Let me try and fix this very quickly!

**Comment by NathanReb** [comment]
Oh no, it's even worse. ocaml-ci runs `dune build @check`, that's highly annoying.

**Comment by mefyl** [comment]
Indeed, even locally I'm constantly walking on eggshells to not have dune fail on me while trying to build those files.

**Comment by NathanReb** [comment]
> Indeed, even locally I'm constantly walking on eggshells to not have dune fail on me while trying to build those files.

I actually have a fix for that where I rewrite the default alias, let me push that at least. I'll look into the CI situation there must be something we can do!

**Comment by NathanReb** [comment]
The point of having library stanzas for those files is so that you can open them in your editor and get your merlin or lsp integration display the error.
That's not as useful now that the compiler reports them nicely so it might just be simpler to remove those stanzas and just keep the expect tests.

**Comment by NathanReb** [comment]
Sorry for the long delay, I'll revive this!

**Comment by NathanReb** [comment]
Can you tick the "Allow maintainers to push" or whatever the name is please? I rebased this and intend to push a couple of extras before merging!

**Comment by NathanReb** [comment]
I just opened #31 to replace this. It's just a rebase and a couple of tiny changes including:
- doc comments to clarify the expected behaviour of some functions in the functor input module type
- rename of some modules and module type
- author and PR ref in the changelog entry

I'm closing this PR, thanks for working on this and sorry again for the long delay!

**Pull request closed without merging**